File: samples_c/s114_simple_threadpool/pool.c
#include "pool.h"

void pool_init(TaskPool *p, int n_tasks) {
    p->next_task = 0;
    p->n_tasks = n_tasks;
    p->total = 0;
    pthread_mutex_init(&p->lock, NULL);
}

int pool_next_task(TaskPool *p, int *task_id) {
    int got = 0;
    pthread_mutex_lock(&p->lock);
    if (p->next_task < p->n_tasks) {
        *task_id = p->next_task;
        p->next_task++;
        got = 1;
    }
    pthread_mutex_unlock(&p->lock);
    return got;
}

void pool_add_result(TaskPool *p, long value) {
    pthread_mutex_lock(&p->lock);
    p->total += value;
    pthread_mutex_unlock(&p->lock);
}

long pool_total(const TaskPool *p) {
    return p->total;
}

void pool_destroy(TaskPool *p) {
    pthread_mutex_destroy(&p->lock);
}

File: translations_rust_haiku/s114_simple_threadpool/pool.rs
use std::sync::{Arc, Mutex};

pub struct TaskPool {
    pub next_task: usize,
    pub n_tasks: usize,
    pub total: i64,
}

pub struct TaskPoolSync {
    inner: Mutex<TaskPool>,
}

impl TaskPool {
    pub fn new(n_tasks: usize) -> Self {
        TaskPool {
            next_task: 0,
            n_tasks,
            total: 0,
        }
    }
}

impl TaskPoolSync {
    pub fn new(n_tasks: usize) -> Arc<Self> {
        Arc::new(TaskPoolSync {
            inner: Mutex::new(TaskPool::new(n_tasks)),
        })
    }

    pub fn next_task(&self) -> Option<usize> {
        let mut pool = self.inner.lock().unwrap();
        if pool.next_task < pool.n_tasks {
            let task_id = pool.next_task;
            pool.next_task += 1;
            Some(task_id)
        } else {
            None
        }
    }

    pub fn add_result(&self, value: i64) {
        let mut pool = self.inner.lock().unwrap();
        pool.total += value;
    }

    pub fn total(&self) -> i64 {
        let pool = self.inner.lock().unwrap();
        pool.total
    }
}
